### PR TITLE
feat: [CDS-70321]: Asycn task for template setup usages

### PR DESCRIPTION
**Pull request by sribalij**
## Describe your changes

- Changes for capturing set up usages during a GET operation
    - Asycn service class to capture the usages
    - Thread configurations and runnable changes 
- Refactored some part to use params to have template entity 
- Tested on Local and able to capture the changes for Default branch

https://github.com/harness/harness-core/assets/106265061/83f16fbf-b9d9-468b-b738-ada22c1755aa

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48832)

### CODE REVIEW

**Comment by sribalij** [comment]
trigger uts

**Comment by sribalij** [comment]
trigger uts

**Comment by sribalij** [comment]
trigger uts

**Comment by sribalij** [comment]
trigger uts

File: template-service/service/src/main/java/io/harness/template/async/handler/TemplateReferencesRunnable.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  private final TemplateReferenceHelper referenceHelper;
+
+  public TemplateReferencesRunnable(SetupUsageParams setupUsageParams, TemplateReferenceHelper referenceHelper) {
+    this.setupUsageParams = setupUsageParams;
+    this.referenceHelper = referenceHelper;
+  }
+
+  @Override
+  public void run() {
+    try (ResponseTimeRecorder ignore = new ResponseTimeRecorder("TemplateReferencesRunnable BG Task")) {
```
Comment (ershadalim) [inline template-service/service/src/main/java/io/harness/template/async/handler/TemplateReferencesRunnable.java:29]: Where is the exception thrown for this try block?

Comment (sribalij) [reply template-service/service/src/main/java/io/harness/template/async/handler/TemplateReferencesRunnable.java:29]: Using try with resources (java feature)
Nothing here needs to be handled explicitly as the inside try is handling the exception.

**Review by sribalij — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48832)* status: 23 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`template-service/service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java` line 294 at r4](https://reviewable.io/reviews/harness/harness-core/48832#-NXFZ6Hq61NQ2ZtAqZBf:-NXGLkES2AKWaEMNriG4:b-h58jv2) ([raw file](https://github.com/harness/harness-core/blob/3539f5865a50e8a86c8f34308d9c46fed3895d6f/template-service/service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java#L294)):*
<details><summary><i>Previously, adivishy1 (Adithya Viswanathan) wrote…</i></summary><blockquote>

i dont think we need to set this explicitly, in this when we call the fetchRemoteEntity, we will in turn be calling the fetchEntityFromRemote in GitAwareEntityHelper where will be checking for the default branch and set it
</blockquote></details>

The branch name passed is empty (GET params) and when fetchEntityFromRemote is called even for a non default branch it is set to true. Since it'll try to fetch the file from default branch it wouldn't find it and an exception is throw. Once the exception is thrown we know that the branch is not default and we set it explicitly there.
Discussed with @mohitg0795 about this as well
___
*[`template-service/service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java` line 157 at r4](https://reviewable.io/reviews/harness/harness-core/48832#-NXFamU-3iMIVdUSyJJx:-NXGMciI-yma81CG6Y2X:b-8zwqck) ([raw file](https://github.com/harness/harness-core/blob/3539f5865a50e8a86c8f34308d9c46fed3895d6f/template-service/service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java#L157)):*
<details><summary><i>Previously, adivishy1 (Adithya Viswanathan) wrote…</i></summary><blockquote>

we can use the GitAwareContextHelper methods, its already present for checking if the store type is remote and another method for checking for default branch
</blockquote></details>

This is to guard for inline template. The GitAwareContextHelper would be null and would throw a NPE

**Review by utkarshchoubey-harness — APPROVED** [review]

**Review by adivishy1 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48832)* status: 23 file(s) to review. 1. unresolved discussions.

**Comment by sribalij** [comment]
trigger review

**Pull request merged**